Sources/SwiftSyntax/Raw/RawSyntaxNodeProtocol.swift

```diff
@@ -225,7 +245,179 @@ public struct RawTokenSyntax: RawSyntaxNodeProtocol {
       leadingTriviaPieces: leadingTriviaPieces,
       trailingTriviaPieces: trailingTriviaPieces,
       presence: .missing,
+      lexerError: nil,
       arena: arena
     )
   }
+
+  /// Assuming that text representing `extendedTrivia` preceeds this token,
+  /// return a token that has its leading trivia prepended by `extendedTrivia`.
+  /// This can be used to transfer trivia from a preceeding token to this token.
+  /// The caller is responsible to delete preceeding trivia from the tree to
+  /// maintain source-fidelity.
+  public func extendingLeadingTrivia(by extendedTrivia: [RawTriviaPiece], arena: SyntaxArena) -> RawTokenSyntax {
+    let extendedTriviaByteLength = extendedTrivia.reduce(0, { $0 + $1.byteLength })
+    switch raw.rawData.payload {
+    case .parsedToken(let dat):
+      assert(
+        String(syntaxText: SyntaxText(baseAddress: dat.wholeText.baseAddress?.advanced(by: -extendedTriviaByteLength), count: extendedTriviaByteLength))
+          == Trivia(pieces: extendedTrivia.map(TriviaPiece.init)).description
+      )
```

Comment:
This is a little scary considering this is `SwiftSyntax` (instead of `Parser`)

This always succeeds in the _current_ parser, but in general "Assuming that text representing `extendedTrivia` preceeds this token," should be more like "Assuming that all `extendedTrivia` text and the entire `self` text are consecutive in a single _buffer_". Maybe: `text` -> `SyntaxText`? Or we might want to have a fallback to `MaterializedToken` path if that's not the case.
Future incremental parsing might break this assertion too, if we introduce more usages of these functions.

Reply:
To clarify, I'm not saying we should support such situation unnecessarily. I just think we need to clarify what is supported.

Reply:
I went with @bnbarham’s suggestion and just create materialized tokens for the re-classified cases. This simplifies parsing and allows us to get rid of all these scary methods. https://github.com/apple/swift-syntax/pull/1255/commits/2deaafd0df2d67440cd41513bccf76d2664d1d19

Reply:
Heh, to be clear my suggestion was to use parsed token and not parse the trivia at all. I don't feel super strongly about this though. Using materialized tokens are a little more heavyweight, I do think it's worth the trade off here. Are you okay with this @rintaro?